### PR TITLE
Remove kubectl from istioctl

### PR DESCRIPTION
**Pull request by ostromart**
This PR refactors the remaining parts of istioctl that were still using kubectl and completely expunges kubectl. All istioctl code will use controller go-client based libraries going forward.

### CODE REVIEW

File: operator/pkg/helmreconciler/wait.go [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+			notReady = append(notReady, "Pod/"+pod.Namespace+"/"+pod.Name)
+		}
+	}
+	return len(notReady) == 0, notReady
+}
+
+func isPodReady(pod *v12.Pod) bool {
+	if len(pod.Status.Conditions) > 0 {
+		for _, condition := range pod.Status.Conditions {
+			if condition.Type == v12.PodReady &&
```
Comment (howardjohn) [inline operator/pkg/helmreconciler/wait.go:238]: nit: v12 is confusing

File: operator/pkg/helmreconciler/wait.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+// WaitForResources polls to get the current status of all pods, PVCs, and Services
+// until all are ready or a timeout is reached
+func WaitForResources(objects object.K8sObjects, restConfig *rest.Config, cs kubernetes.Interface, waitTimeout time.Duration, dryRun bool, l clog.Logger) error {
+	if dryRun {
+		l.LogAndPrint("Not waiting for resources ready in dry run mode.")
+		return nil
+	}
+
+	if err := waitForCRDs(objects, restConfig); err != nil {
```
Comment (howardjohn) [inline operator/pkg/helmreconciler/wait.go:48]: nit: can this use `cs` instead of `restConfig`?

File: operator/pkg/helmreconciler/render.go [hunk 3]
```diff
@@ -386,6 +367,71 @@ func toChartManifestsMap(m name.ManifestMap) ChartManifestsMap {
...
+	}
+	objAccessor, err := meta.Accessor(h.iop)
+	if err != nil {
+		return "", err
+	}
+	return objAccessor.GetName(), nil
+}
+
+func (h *HelmReconciler) getCRHash(componentName string) (string, error) {
+	crName, err := h.getCRName()
```
Comment (howardjohn) [inline operator/pkg/helmreconciler/render.go:428]: name is not unique -  should we have namespace here?

File: operator/pkg/helmreconciler/render.go [hunk 4]
```diff
@@ -386,6 +367,71 @@ func toChartManifestsMap(m name.ManifestMap) ChartManifestsMap {
...
+	}
+
+	componentLabelValue := componentName
+	// Only pilot component uses revisions
+	if componentName == string(name.PilotComponentName) {
+		if revision == "" {
+			revision = "default"
+		}
+		labels[model.RevisionLabel] = revision
+		componentLabelValue += "-" + revision
```
Comment (howardjohn) [inline operator/pkg/helmreconciler/render.go:389]: the `kubectl` model only uses this for `-l` for the prune -- not adding the label

File: operator/cmd/mesh/operator-common.go [hunk 5]
```diff
@@ -82,3 +94,113 @@ tag: {{.Tag}}
...
 	return vals, manifest, err
 }
+
+var k8sRESTConfig *rest.Config
+
+var k8sClientset *kubernetes.Clientset
+
+var currentKubeconfig string
+
+var currentContext string
```
Comment (howardjohn) [inline operator/cmd/mesh/operator-common.go:104]: can we remove these? can be in another pr

**Review by howardjohn — COMMENTED** [review]
lgtm mostly

File: operator/pkg/helmreconciler/wait.go (hunk 2 above)
Comment (ostromart) [reply operator/pkg/helmreconciler/wait.go:48]: i don't think so because waitForCRDs needs an apiextensions client.

File: operator/pkg/helmreconciler/wait.go (hunk 1 above)
Comment (ostromart) [reply operator/pkg/helmreconciler/wait.go:238]: done.

File: operator/pkg/helmreconciler/render.go (hunk 3 above)
Comment (ostromart) [reply operator/pkg/helmreconciler/render.go:428]: yes, good point.

File: operator/pkg/helmreconciler/render.go (hunk 4 above)
Comment (ostromart) [reply operator/pkg/helmreconciler/render.go:389]: done

File: operator/cmd/mesh/operator-common.go [hunk 6]
```diff
@@ -15,8 +15,21 @@
 package mesh
 
 import (
+	"context"
+	"fmt"
+	"os" // Import all Kubernetes client auth plugins (e.g. Azure, GCP, OIDC, etc.)
```
Comment (howardjohn) [inline operator/cmd/mesh/operator-common.go:20]: nit: wrong line for comment

File: operator/cmd/mesh/operator-remove.go [hunk 7]
```diff
@@ -101,32 +100,7 @@ func operatorRemove(args *rootArgs, orArgs *operatorRemoveArgs, l clog.Logger, d
...
-	}
-
-	if !ignoreError(stderr) {
-		l.LogAndPrint("Error detail:\n", stderr, "\n")
-		l.LogAndPrint(stdout, "\n")
-		success = false
-	}
-	return success
+func deleteManifest(_, _ string, _ *Options, l clog.Logger) bool {
+	l.LogAndError("Deleting manifest not implemented")
```
Comment (howardjohn) [inline operator/cmd/mesh/operator-remove.go:104]: if its not implemented should we just delete the function?

**Review by howardjohn — COMMENTED** [review]
lgtm if tests pass

File: go.sum [hunk 8]
```diff
@@ -988,6 +988,7 @@ istio.io/api v0.0.0-20200417223136-90a960729620/go.mod h1:kyq3g5w42zl/AKlbzDGppY
 istio.io/gogo-genproto v0.0.0-20190930162913-45029607206a/go.mod h1:OzpAts7jljZceG4Vqi5/zXy/pOg1b209T3jb7Nv5wIs=
 istio.io/gogo-genproto v0.0.0-20200326154102-997c228eecef h1:jVlVwrFW1LIm9XyPnvthNEvLjALn1zsTysGRKctl4Lc=
 istio.io/gogo-genproto v0.0.0-20200326154102-997c228eecef/go.mod h1:OzpAts7jljZceG4Vqi5/zXy/pOg1b209T3jb7Nv5wIs=
+istio.io/operator v1.3.0 h1:TLBbQk00U6XjoO24sFvT8oRgtYmRfKATBEnker2gnNY=
```
[the remark this answers is not in the document]
Comment (ostromart) [reply go.sum]: done.

File: operator/cmd/mesh/operator-common.go (hunk 5 above)
Comment (ostromart) [reply operator/cmd/mesh/operator-common.go]: done.

File: operator/cmd/mesh/operator-remove.go (hunk 7 above)
Comment (ostromart) [reply operator/cmd/mesh/operator-remove.go]: this is todo, i didn't want to tackle in this pr but it's needed.

File: operator/cmd/mesh/operator-common.go (hunk 6 above)
Comment (ostromart) [reply operator/cmd/mesh/operator-common.go]: done.

File: operator/pkg/helmreconciler/render.go (hunk 3 above)
Comment (ostromart) [reply operator/pkg/helmreconciler/render.go]: done.

**Review by howardjohn — APPROVED** [review]

**Review by richardwxn — APPROVED** [review]

File: operator/cmd/mesh/upgrade.go [hunk 9]
```diff
@@ -202,31 +205,13 @@ func upgrade(rootArgs *rootArgs, args *upgradeArgs, l clog.Logger) (err error) {
...
 	waitForConfirmation(args.skipConfirmation, l)
 
-	// Run pre-upgrade hooks
-	hparams := &hooks.HookCommonParams{
-		SourceVer:  currentVersion,
-		TargetVer:  targetVersion,
-		SourceIOPS: targetIOPS,
-		TargetIOPS: targetIOPS,
-	}
-	errs := hooks.RunPreUpgradeHooks(kubeClient, hparams, rootArgs.dryRun)
```
[the remark this answers is not in the document]
Comment (ostromart) [reply operator/cmd/mesh/upgrade.go:212]: they weren't used and depended on some of the old code so i just took this opportunity to remove them. i think if we have a use case for the hooks we could re-implement using this older implementation for inspiration.

File: operator/cmd/mesh/shared.go [hunk 10]
```diff
@@ -102,3 +112,94 @@ func confirm(msg string, writer io.Writer) bool {
...
+	b, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return "", "", fmt.Errorf("could not read values from file %s: %s", filePath, err)
+	}
+	customResource = string(b)
+	istioNamespace = v1alpha1.Namespace(mergedIOPS)
+	return
+}
+
+func genNamespaceResource(namespace string) string {
```
[the remark this answers is not in the document]
Comment (ostromart) [reply operator/cmd/mesh/shared.go:185]: thanks for pointing out. i made the change but pr had already merged so i'll put in the next one.
